File: cli-python/src/code_archaeology/annotator/templates.py
from __future__ import annotations

from ..types import ArchaeologyResult


def generate_brief(result: ArchaeologyResult) -> list[str]:
    span = _get_span_text(result)
    lines = [
        "[Code Archaeology]",
        f"History: {result.total_changes} changes over {span} ({len(result.contributors)} contributors)",
    ]
    if result.timeline:
        latest = result.timeline[0]
        date = latest.commit.date.strftime("%Y-%m-%d")
        pr = f" (#{latest.pull_request.number})" if latest.pull_request else ""
        lines.append(f"Last major change: {latest.commit.message}{pr} - {date}")
    return lines


def generate_detailed(result: ArchaeologyResult) -> list[str]:
    span = _get_span_text(result)
    lines = [
        "[Code Archaeology]",
        f"History: {result.total_changes} changes over {span} ({len(result.contributors)} contributors)",
    ]
    if result.timeline:
        latest = result.timeline[0]
        date = latest.commit.date.strftime("%Y-%m-%d")
        pr = f" (#{latest.pull_request.number})" if latest.pull_request else ""
        lines.append(f"Last major change: {latest.commit.message}{pr} - {date}")

    lines.append("Key decisions:")
    for entry in result.timeline[:5]:
        date = entry.commit.date.strftime("%Y-%m-%d")
        pr = f" (#{entry.pull_request.number})" if entry.pull_request else ""
        lines.append(f"  - {entry.commit.message}{pr} ({date})")

    if len(result.contributors) > 1:
        lines.append(f"Contributors: {', '.join(result.contributors)}")

    return lines


def generate_annotation(result: ArchaeologyResult, style: str = "detailed") -> list[str]:
    if style == "brief":
        return generate_brief(result)
    return generate_detailed(result)


def _get_span_text(result: ArchaeologyResult) -> str:
    if not result.time_span_first or not result.time_span_last:
        return "unknown"
    days = (result.time_span_last - result.time_span_first).days
    if days < 1:
        return "less than a day"
    if days < 30:
        return f"{days} days"
    months = days // 30
    if months < 12:
        return f"{months} months"
    years = months // 12
    rem = months % 12
    return f"{years}y {rem}m" if rem else f"{years} years"

File: cli-node/src/annotator/templates.ts
import { ArchaeologyResult, AnnotationOptions } from '../types.js';

export function generateBriefAnnotation(result: ArchaeologyResult): string[] {
  const span = getSpanText(result.timeSpan.first, result.timeSpan.last);
  const latest = result.timeline[0];
  const lines: string[] = [
    '[Code Archaeology]',
    `History: ${result.totalChanges} changes over ${span} (${result.contributors.length} contributors)`,
  ];

  if (latest) {
    const date = latest.commit.date.toISOString().slice(0, 10);
    const pr = latest.pullRequest ? ` (#${latest.pullRequest.number})` : '';
    lines.push(`Last major change: ${latest.commit.message}${pr} - ${date}`);
  }

  return lines;
}

export function generateDetailedAnnotation(result: ArchaeologyResult): string[] {
  const span = getSpanText(result.timeSpan.first, result.timeSpan.last);
  const lines: string[] = [
    '[Code Archaeology]',
    `History: ${result.totalChanges} changes over ${span} (${result.contributors.length} contributors)`,
  ];

  if (result.timeline.length > 0) {
    const latest = result.timeline[0];
    const date = latest.commit.date.toISOString().slice(0, 10);
    const pr = latest.pullRequest ? ` (#${latest.pullRequest.number})` : '';
    lines.push(`Last major change: ${latest.commit.message}${pr} - ${date}`);
  }

  lines.push('Key decisions:');
  for (const entry of result.timeline.slice(0, 5)) {
    const date = entry.commit.date.toISOString().slice(0, 10);
    const pr = entry.pullRequest ? ` (#${entry.pullRequest.number})` : '';
    lines.push(`  - ${entry.commit.message}${pr} (${date})`);
  }

  if (result.contributors.length > 1) {
    lines.push(`Contributors: ${result.contributors.join(', ')}`);
  }

  return lines;
}

export function generateAnnotation(result: ArchaeologyResult, options: AnnotationOptions): string[] {
  if (options.style === 'brief') {
    return generateBriefAnnotation(result);
  }
  return generateDetailedAnnotation(result);
}

function getSpanText(first: Date, last: Date): string {
  const diffMs = last.getTime() - first.getTime();
  const days = Math.floor(diffMs / (1000 * 60 * 60 * 24));
  if (days < 1) return 'less than a day';
  if (days < 30) return `${days} days`;
  const months = Math.floor(days / 30);
  if (months < 12) return `${months} months`;
  const years = Math.floor(months / 12);
  const rem = months % 12;
  return rem > 0 ? `${years}y ${rem}m` : `${years} years`;
}
